Migration guide - Upgrade Node.js to v18 and npm to v9 (#1854)
spryker/spryker-docs: docs/scos/dev/front-end-development/202304.0/migration-guide-upgrade-nodejs-to-v18.md

*Estimated migration time: 1h*

## 1) Update configuration files

1. Set up new versions of `node/npm` in the main `*.yml` files, like `deploy.yml`, `deploy.dev.yml`, and `deploy.ci.yml`:

```yaml
image:
    ...
    node:
        version: 18
        npm: 9
```

{% info_block infoBox "Note" %}

To ensure the CI jobs run successfully, add the same config part to all `deploy.*.yml` files used by the frontend.

{% endinfo_block %}

2. In the root directory, create/update `.nvmrc` file with the following content:

```text
18.16.0
```

3. Update engines and dependencies in the `package.json`:

```json
{
    ...
    "engines": {
        "node": ">=18.0.0",
        "npm": ">=9.0.0"
    },
    ...
    "devDependencies": {
        ...
        "@spryker/oryx-for-zed": "~3.0.0",
        "autoprefixer": "~10.4.14",
        "postcss": "~8.4.23",
        "postcss-loader": "~7.3.0",
        "sass-resources-loader": "~2.2.5",
        ...
    },
    ...
}
```

{% info_block infoBox "Note" %}

Keep in mind *@spryker/sass-resources-loader* must be replaced with *sass-resources-loader* dependency.

> {% endinfo_block %}

Reviewer: Please add it as a list item with `Remove` title e.g.:
Remove the `"@spryker/sass-resources-loader": "x.x.x"` dependency.

4. Update webpack config in `frontend/configs/development.js`:

```js
webpack: {
    module: {
        rules: [
            ...,
            {
                test: /\.scss/i,
                use: [
                    ...,
                    {
                        loader: 'postcss-loader',
                        options: {
                            postcssOptions: {
                                plugins: [

>                                     require('autoprefixer')

Reviewer: Is not clear what should be updated.
Please add parts with should be replaced.

---
title: Migration guide - Upgrade Node.js to v18 and npm to v9
last_updated: May 16, 2023
description: Use this guide to upgrade Node.js to v18 and npm to v9.
template: concept-topic-template
---

This document provides instructions for upgrading Node.js to v18 and npm to v9.

## Overview

According to [Node.js schedule](https://github.com/nodejs/release#release-schedule) the 16 version will be maintained till 11.09.2023. So Spryker recommends migrating to 18 LTS version.

*Estimated migration time: 1h*

## 1) Update configuration files

1. Set up new versions of `node/npm` in the main `*.yml` files, like `deploy.yml`, `deploy.dev.yml`, and `deploy.ci.yml`:

```yaml
image:
    ...
    node:
        version: 18
        npm: 9
```

{% info_block infoBox "Note" %}

To ensure the CI jobs run successfully, add the same config part to all `deploy.*.yml` files used by the frontend.

{% endinfo_block %}

2. In the root directory, create/update `.nvmrc` file with the following content:

```text
18.16.0
```

> 3. Update engines and dependencies in the `package.json`:

Reviewer: ```suggestion
3. Update or add dependencies and engines in the `package.json`:
```

    ...
    "devDependencies": {
        ...,
        "@babel/preset-env": "~7.20.2",
        "@spryker/oryx-for-zed": "~3.0.0",
        "autoprefixer": "~10.4.14",
        "postcss": "~8.4.24",
        "postcss-loader": "~7.3.0",
        "sass-resources-loader": "~2.2.5",
        ...
    },
    ...
}
```

4. Remove dependency in the `package.json`:

```json
    "@spryker/sass-resources-loader": "x.x.x"
```

5. Update webpack config in `frontend/configs/development.js`:
  - Replace `options` of `postcss-loader`;
  - Replace `@spryker/sass-resources-loader` with `sass-resources-loader`;

```js
    ...
        {
            loader: 'postcss-loader',
            options: {
                postcssOptions: {
                    plugins: [
                        require('autoprefixer')
                    ]
                }
            }
        },
        ...,
        {
            loader: 'sass-resources-loader',
            options: {
                resources: [sharedScss, ...styles],
            },
        },
    ...
```

## 2) Build the project

1. Apply the docker changes:

```bash
docker/sdk boot deploy.dev.yml
docker/sdk up
```

2. Regenerate `package-lock.json`:

```bash

> rm -rf package-lock.json

Reviewer: We shouldn't do this by default, it can be as a note if they face some issues.

---
title: Migration guide - Upgrade Node.js to v18 and npm to v9
last_updated: May 16, 2023
description: Use this guide to upgrade Node.js to v18 and npm to v9.
template: concept-topic-template
---

This document provides instructions for upgrading Node.js to v18 and npm to v9.

## Overview

> According to [Node.js schedule](https://github.com/nodejs/release#release-schedule) the 16 version will be maintained till 11.09.2023. So Spryker recommends migrating to 18 LTS version.

Reviewer 2: ```suggestion
According to [Node.js schedule](https://github.com/nodejs/release#release-schedule), the maintenance of version 16  stops on September 11, 2023. Therefore, we recommend migrating to the 18 LTS version.
```

*Estimated migration time: 1h*

## 1) Update configuration files

> 1. Set up new versions of `node/npm` in the main `*.yml` files, like `deploy.yml`, `deploy.dev.yml`, and `deploy.ci.yml`:

Reviewer 2: ```suggestion
1. In the main `*.yml` files, set up new versions of `node/npm`, like `deploy.yml`, `deploy.dev.yml`, and `deploy.ci.yml`:
```

```yaml
image:
    ...
    node:
        version: 18
        npm: 9
```

{% info_block infoBox "Note" %}

To ensure the CI jobs run successfully, add the same config part to all `deploy.*.yml` files used by the frontend.

{% endinfo_block %}

> 2. In the root directory, create/update `.nvmrc` file with the following content:

Reviewer 2: ```suggestion
2. In the root directory, create or update the `.nvmrc` file with the following content:
```

```text
18.16.0
```

> 3. Update or add dependencies and engines in the `package.json`:

Reviewer 2: ```suggestion
3. In the `package.json`, update or add dependencies and engines:
```

```json
{
    ...
    "engines": {
        "node": ">=18.0.0",
        "npm": ">=9.0.0"
    },
    ...
    "devDependencies": {
        ...,
        "@babel/preset-env": "~7.20.2",
        "@spryker/oryx-for-zed": "~3.0.0",
        "autoprefixer": "~10.4.14",
        "postcss": "~8.4.24",
        "postcss-loader": "~7.3.0",
        "sass-resources-loader": "~2.2.5",
        ...
    },
    ...
}
```

> 4. Remove dependency in the `package.json`:

Reviewer 2: ```suggestion
4. In the `package.json`, remove the dependency :
```

```json
    "@spryker/sass-resources-loader": "x.x.x"
```

> 5. Update webpack config in `frontend/configs/development.js`:

Reviewer 2: ```suggestion
5. In `frontend/configs/development.js`, update the webpack config:
```

>   - Replace `options` of `postcss-loader`;

Reviewer 2: If these are steps, then use the numbered list instead of bulleted points. Also, correct grammar:
```suggestion
  1. In `postcss-loader`, replace `options`.
```

  - Replace `@spryker/sass-resources-loader` with `sass-resources-loader`;

```js
    ...
        {
            loader: 'postcss-loader',
            options: {
                postcssOptions: {
                    plugins: [
                        require('autoprefixer')
                    ]
                }
            }
        },
        ...,
        {
            loader: 'sass-resources-loader',
            options: {
                resources: [sharedScss, ...styles],
            },
        },
    ...
```

## 2) Build the project

1. Apply the docker changes:

```bash
docker/sdk boot deploy.dev.yml
docker/sdk up
```

2. Regenerate `package-lock.json`:

```bash
docker/sdk cli npm install
```

{% info_block infoBox "Note" %}

> Ensure that the `package-lock.json` file has `"lockfileVersion": 2` or higher, otherwise need to remove `node_modules` and `package-lock.json`. After that regenerate again.

Reviewer 2: Make sure I rephrased the sentence correctly, and that it carries the same meaning.
```suggestion
Ensure that the `package-lock.json` file has `"lockfileVersion": 2` or later; otherwise, remove `node_modules` and `package-lock.json` and regenerate the file again.
```

```bash
rm -rf node_modules && docker/sdk cli rm -rf node_modules && rm -rf package-lock.json
docker/sdk cli npm install
```

{% endinfo_block %}

3. Build the project using Node.js v18 and npm v9:

```bash
rm -rf node_modules && docker/sdk cli rm -rf node_modules
docker/sdk up --build --assets --data
```

{% info_block infoBox "Note" %}

If you use `Node.js` and `npm` locally, make sure their versions are correct:

```bash
node -v
npm -v
```

To update the versions, use official documentation:

- [Node.js](https://nodejs.org/en/download/package-manager)
- [npm](https://docs.npmjs.com/try-the-latest-stable-version-of-npm)

{% endinfo_block %}

## 3) Update GitHub Actions

> In case using GitHub Actions on the project need to update `.github/workflows/ci.yml` file.

Reviewer 2: Should it be step 1?
```suggestion
In case using GitHub Actions on the project, update the `.github/workflows/ci.yml` file.
```

> Update Node.js version:

Reviewer 2: Should it be step 2?

```yaml
- uses: actions/setup-node@v3
    with:
      node-version: '18'
```

> Remove npm cache:

Reviewer 2: Should it be step 3?

            options: {
                resources: [sharedScss, ...styles],
            },
        },
    ...
```

## 2) Build the project

1. Apply the docker changes:

```bash
docker/sdk boot deploy.dev.yml
docker/sdk up
```

2. Regenerate `package-lock.json`:

```bash
docker/sdk cli npm install
```

{% info_block infoBox "Note" %}

Ensure that the `package-lock.json` file has `"lockfileVersion": 2` or later; otherwise, remove `node_modules` and `package-lock.json` and regenerate the file again.

```bash
rm -rf node_modules && docker/sdk cli rm -rf node_modules && rm -rf package-lock.json
docker/sdk cli npm install
```

{% endinfo_block %}

3. Build the project using Node.js v18 and npm v9:

```bash
rm -rf node_modules && docker/sdk cli rm -rf node_modules
docker/sdk up --build --assets --data
```

{% info_block infoBox "Note" %}

If you use `Node.js` and `npm` locally, make sure their versions are correct:

```bash
node -v
npm -v
```

To update the versions, use official documentation:

- [Node.js](https://nodejs.org/en/download/package-manager)
- [npm](https://docs.npmjs.com/try-the-latest-stable-version-of-npm)

{% endinfo_block %}

## 3) Update GitHub Actions

> If you are using GitHub Actions on the project, update the `.github/workflows/ci.yml` file:

Reviewer: ```suggestion
If the project uses CI, adjust .github/workflows/ci.yml:
```
It's already used [here](https://docs.spryker.com/docs/scos/dev/front-end-development/202212.0/migration-guide-switch-from-tslint-to-eslint.html#create-or-update-configuration-files), so it should be consistent.
